### PR TITLE
Add extendedwatercannons special world property

### LINKED ISSUE #3315: Only 3 water cannons on Fire Truck can be used at the same time
### Describe the bug

idk if this was intended but only 3 water cannons are spraying out the water
image

### Steps to reproduce

1. Spawn 5 Fire Trucks
2. Warp 5 peds into each Fire Trucks
3. Use setPedControlState(ped, "vehicle_fire", true)
4. Only 3 water cannons are spraying out the water

Or use this resource:
[watercannon.zip](https://github.com/multitheftauto/mtasa-blue/files/14298367/watercannon.zip)
1. Goto coordinate 0, 0, 0,
2. Start the resource

### Version

Client: Multi Theft Auto v1.6-release-22388 (Windows)
Server: MTA:SA Server v1.6-release-22388 (Windows)

### Additional context

_No response_

### Relevant log output

_No response_

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not security related.

### PR DESCRIPTION
**Pull request by FileEX**
Added ``extendedwatercannons`` special world property, that allows the maximum number of water cannons to be extended at the same time. I set the new limit to 30 because that number seems to be enough, perhaps even too much. 

image

Thanks to @Merlin for help with crashes

Resolved #3315

### CODE REVIEW

**Comment by Dark-Dragon** [comment]
Would there be any disadvantage to just increasing this limit by default? I don't think this is a case where the limited default behavior is desirable at all, therefore being able to toggle between the two seems rather unnecessary to me.

**Comment by FileEX** [comment]
> Would there be any disadvantage to just increasing this limit by default? I don't think this is a case where the limited default behavior is desirable at all, therefore being able to toggle between the two seems rather unnecessary to me.

The same can be said about vegetation (``randomfoliage``) and underwater creatures - toggle them seems unnecessary, but you can toggle them anyway. I believe we can set this property to enabled by default, but I still think leaving the option to disable it isn't a bad thing. It is better when the user has choice and freedom than predefined settings that he cannot change.

In addition, it seems to me that few people really need an increased limit. In 12 years of playing, I never even noticed that such a limit even existed, and no one had ever reported it before.

**Comment by Dark-Dragon** [comment]
Well there could be had a conversation about what the reasons might be it wasn't reported in the first place, but having it on by default seems like a fine compromise. Still I wouldn't be surprised if it will take more than 12 years for someone to specifically need to disable it, whereas I can see scenarios where foliage and water creatures could be desired or undesired. I'm just thinking the special world properties are slowly starting to get rather crowded, but that's not for me to decide, so I just want to point it out whenever I can.

**Comment by FileEX** [comment]
I think it's good that we get new world properties. Thanks to this, server creators on MTA have more and more freedom and more opportunities. If someone doesn't want foliages, they can disable it, if someone doesn't want flipped cars to burn, they can disable this property. The same in this case, if for some reason someone really does not need such a limit extension, then disable this property. I think it's better to have this option than not to have it.

**Comment by Lpsd** [comment]
Please fix conflicts and we can merge this

**Comment by FileEX** [comment]
> Please fix conflicts and we can merge this

Done

**Comment by lopezloo** [comment]
I think this limit should be increased by default.

**Comment by FileEX** [comment]
> I think this limit should be increased by default.

As you can see https://github.com/multitheftauto/mtasa-blue/pull/3365/commits/bff7007fe674e6c75f129f1ac68b39ed9a2a2d05

The limit is increased by default

File: Client/game_sa/CGameSA.h [hunk 1]
```diff
@@ -347,6 +364,7 @@ class CGameSA : public CGame
     bool         m_areWaterCreaturesEnabled{true};
     bool         m_isBurnFlippedCarsEnabled{true};
     bool         m_isFireballDestructEnabled{true};
+    bool         m_isExtendedWaterCannonsEnabled;
```
Comment (botder) [inline Client/game_sa/CGameSA.h]:
```suggestion
    bool         m_isExtendedWaterCannonsEnabled{false};
```
You were using the default initializer, which uses a random value at whatever address this member ends up at. In a Debug build this is probably set to `false` anyway, so you would never notice this.

File: Client/game_sa/CGameSA.h [hunk 2]
```diff
@@ -82,6 +82,20 @@ extern unsigned int OBJECTDYNAMICINFO_MAX;            // default: 160
 #define CHEAT_NEVERWANTED           "neverwanted"
 #define CHEAT_HEALTARMORMONEY       "healtharmormoney"
 
+#define FUNC_CAESoundManager_CancelSoundsOwnedByAudioEntity 0x4EFCD0
+#define STRUC_CAESoundManager                               0xB62CB0
```
Comment (botder) [inline Client/game_sa/CGameSA.h]: Shouldn't this be `STRUCT_`?

File: Shared/sdk/net/bitstream.h [hunk 3]
```diff
@@ -540,8 +540,17 @@ enum class eBitStreamVersion : unsigned short
     // 2023-10-12
     CPlayerJoinCompletePacket_ServerName,
 
+
+    // Add "extendedwatercannons" to setWorldSpecialPropertyEnabled
+    // 2024-05-23
+    WorldSpecialProperty_ExtendedWaterCannons,
```
Comment (botder) [inline Shared/sdk/net/bitstream.h:549]: This should be added at the end of the enum.

File: Shared/sdk/net/bitstream.h [hunk 4]
```diff
@@ -540,8 +540,17 @@ enum class eBitStreamVersion : unsigned short
...
+    
+    // Send weapon switch for other players
+    // 2023-01-13
+    OnPlayerWeaponSwitch_Remote,
+
     // Add "roadsignstext" to setWorldSpecialPropertyEnabled
-    // 2024-05-17
+    // 2024-05-23
```
Comment (botder) [inline Shared/sdk/net/bitstream.h]: Do not modify the date of this enum field.

File: Shared/sdk/net/SyncStructures.h [hunk 5]
```diff
@@ -2065,6 +2069,7 @@ struct SWorldSpecialPropertiesStateSync : public ISyncStructure
         if (bitStream.Can(eBitStreamVersion::WorldSpecialProperty_FireballDestruct))
             bitStream.WriteBits(reinterpret_cast<const char*>(&data2), BITCOUNT2);
 
+        if (bitStream.Can(eBitStreamVersion::WorldSpecialProperty_ExtendedWaterCannons))
         if (bitStream.Can(eBitStreamVersion::WorldSpecialProperty_RoadSignsText))
             bitStream.WriteBits(reinterpret_cast<const char*>(&data3), BITCOUNT3);
```
Comment (botder) [inline Shared/sdk/net/SyncStructures.h]: Merge error? Looks like you're missing a WriteBits call for the first if-statement.

File: Shared/sdk/net/SyncStructures.h [hunk 6]
```diff
@@ -2097,6 +2102,7 @@ struct SWorldSpecialPropertiesStateSync : public ISyncStructure
 
     struct
     {
+        bool extendedwatercannons : 1;
         bool roadsignstext : 1;
```
Comment (botder) [inline Shared/sdk/net/SyncStructures.h]: Your new bit should be in a new struct `data4`.

File: Shared/sdk/net/SyncStructures.h [hunk 7]
```diff
@@ -2034,7 +2034,7 @@ struct SWorldSpecialPropertiesStateSync : public ISyncStructure
     };
     enum
     {
-        BITCOUNT3 = 1
+        BITCOUNT3 = 2
```
Comment (botder) [inline Shared/sdk/net/SyncStructures.h]: You must use `BITCOUNT4`, don't change `BITCOUNT3` at all.

File: Shared/mods/deathmatch/logic/Enums.h [hunk 8]
```diff
@@ -88,6 +88,7 @@ enum class WorldSpecialProperty
     WATERCREATURES,
     BURNFLIPPEDCARS,
     FIREBALLDESTRUCT,
+    EXTENDEDWATERCANNONS,
     ROADSIGNSTEXT,
```
Comment (botder) [inline Shared/mods/deathmatch/logic/Enums.h]: Move your new field to the bottom of the list.

**Review by botder — CHANGES_REQUESTED** [review]

File: Shared/sdk/net/SyncStructures.h (hunk 5 above)
Comment (FileEX) [reply Shared/sdk/net/SyncStructures.h]: Some strange things happened during conflict resolution. I've already fixed everything

File: Shared/sdk/net/bitstream.h [hunk 9]
```diff
@@ -539,11 +539,19 @@ enum class eBitStreamVersion : unsigned short
     // Send server name to player in CPlayerJoinCompletePacket
     // 2023-10-12
     CPlayerJoinCompletePacket_ServerName,
+    
+    // Send weapon switch for other players
+    // 2023-01-13
+    OnPlayerWeaponSwitch_Remote,
```
Comment (botder) [inline Shared/sdk/net/bitstream.h]: This does not belong here

**Review by botder — CHANGES_REQUESTED** [review]

File: Shared/sdk/net/bitstream.h (hunk 9 above)
Comment (FileEX) [reply Shared/sdk/net/bitstream.h]: Fixed

**Review by botder — APPROVED** [review]

**Pull request merged**